File: Controle Quantidade Lançamentos/Delphi/Unit1.pas
unit Unit1;

interface

uses
  Winapi.Windows, Winapi.Messages, System.SysUtils, System.Variants, System.Classes, Vcl.Graphics,
  Vcl.Controls, Vcl.Forms, Vcl.Dialogs, Vcl.ExtCtrls, Vcl.StdCtrls,
  Vcl.ComCtrls, Data.DBXMySQL, Data.DB, Data.SqlExpr, Vcl.Grids, Vcl.DBGrids, Data.FMTBcd,
  Datasnap.Provider, Datasnap.DBClient, DateUtils;

type
  TForm1 = class(TForm)
    Panel1: TPanel;
    PageControl1: TPageControl;
    TabSheet1: TTabSheet;
    Conexao: TSQLConnection;
    Button1: TButton;
    DBGrid1: TDBGrid;
    EditMes: TLabeledEdit;
    EditAno: TLabeledEdit;
    EditVencimento: TLabeledEdit;
    EditBloqueio: TLabeledEdit;
    EditLiberado: TLabeledEdit;
    Button2: TButton;
    SQLTable1: TSQLTable;
    DataSource1: TDataSource;
    ClientDataSet1: TClientDataSet;
    DataSetProvider1: TDataSetProvider;
    SQLTable1id: TIntegerField;
    SQLTable1a: TStringField;
    SQLTable1b: TStringField;
    SQLTable1c: TStringField;
    SQLTable1d: TStringField;
    SQLTable1e: TStringField;
    ClientDataSet1id: TIntegerField;
    ClientDataSet1a: TStringField;
    ClientDataSet1b: TStringField;
    ClientDataSet1c: TStringField;
    ClientDataSet1d: TStringField;
    ClientDataSet1e: TStringField;
    SQLTable1f: TStringField;
    ClientDataSet1f: TStringField;
    EditNomeTabela: TLabeledEdit;
    Label1: TLabel;
    DBGrid2: TDBGrid;
    DataSource2: TDataSource;
    ClientDataSet2: TClientDataSet;
    ClientDataSet2id: TIntegerField;
    ClientDataSet2a: TStringField;
    ClientDataSet2b: TStringField;
    procedure Button1Click(Sender: TObject);
    procedure Button2Click(Sender: TObject);
  private
    { Private declarations }
  public
    { Public declarations }
  end;

var
  Form1: TForm1;

implementation

{$R *.dfm}

uses Biblioteca;

// Exercício: Criptografe os dados na tabela conforme exemplo anterior
// Exercício: Implemente o controle num arquivo fake conforme exemplo anterior

{$Region 'Controle via Banco'}
procedure TForm1.Button1Click(Sender: TObject);
var
  Achou: Boolean;
begin
  Achou := False;

  ClientDataSet2.First;
  while not ClientDataSet2.eof do
  begin
    if ClientDataSet2a.AsString = EditNomeTabela.Text then
    begin
      Achou := True;

      if ClientDataSet1.RecordCount <= ClientDataSet2b.AsInteger then
      begin
        ClientDataSet1.Open;
        //
        ClientDataSet1.Append;
        ClientDataSet1.FieldByName('a').AsString := Codifica('C', EditMes.Text);
        ClientDataSet1.FieldByName('b').AsString := Codifica('C', EditAno.Text);
        ClientDataSet1.FieldByName('c').AsString := Codifica('C', EditVencimento.Text);
        ClientDataSet1.FieldByName('d').AsString := Codifica('C', EditBloqueio.Text);
        ClientDataSet1.FieldByName('e').AsString := Codifica('C', EditLiberado.Text);

        ClientDataSet1.FieldByName('f').AsString :=
                                                    MD5String(
                                                    ClientDataSet1.FieldByName('a').AsString +
                                                    ClientDataSet1.FieldByName('b').AsString +
                                                    ClientDataSet1.FieldByName('c').AsString +
                                                    ClientDataSet1.FieldByName('d').AsString +
                                                    ClientDataSet1.FieldByName('e').AsString
                                                    );
        ClientDataSet1.Post;
        ClientDataSet1.ApplyUpdates(0);
        //
        SQLTable1.Close;
        SQLTable1.Open;
        ClientDataSet1.Close;
        ClientDataSet1.Open;
      end
      else
        ShowMessage('Não é possível realizar mais lançamentos nessa tabela.')

    end;

    if Achou then
      Break;

    ClientDataSet2.Next;
  end;

end;

procedure TForm1.Button2Click(Sender: TObject);
begin
  SQLTable1.Open;
  //
  ClientDataSet1.Close;
  ClientDataSet1.Open;
end;

{$EndRegion 'Controle via Banco'}

end.

File: Controle Quantidade Lançamentos/C#/Unit1.cs
using MySql.Data.MySqlClient;
using System;
using System.Collections.Generic;
using System.Windows.Forms;
using Util;
using VO;

namespace Seguranca
{

    // Exercício: Criptografe os dados na tabela conforme exemplo anterior
    // Exercício: Implemente o controle num arquivo fake conforme exemplo anterior

	  
	public partial class Form1 : Form
	{

        private static List<ControlePagamentoVO> ListaPagamento;

        private MySqlConnection conexao;
        private MySqlCommand comando;
        private MySqlDataReader leitor;
        public string ConsultaSQL;
		
		public Form1()
		{
			// Required for Windows Form Designer support
			InitializeComponent();
            Grid1.AutoGenerateColumns = false;
        }

        private void Button1_Click(object sender, EventArgs e)
        {
            if (ListaPagamento.Count <= 5)
            {
                ConsultaSQL = "insert into controle_pagamento ( " +
                                                        "a,b,c,d,e,f)" +
                                                        " values (" +
                                                        "?pa,?pb,?pc,?pd,?pe,?pf)";
                try
                {
                    conexao = dbConnection.conectar();
                    comando = new MySqlCommand(ConsultaSQL, conexao);

                    comando.Parameters.AddWithValue("?pa", Biblioteca.Encripta(EditMes.Text));
                    comando.Parameters.AddWithValue("?pb", Biblioteca.Encripta(EditAno.Text));
                    comando.Parameters.AddWithValue("?pc", Biblioteca.Encripta(EditVencimento.Text));
                    comando.Parameters.AddWithValue("?pd", Biblioteca.Encripta(EditBloqueio.Text));
                    comando.Parameters.AddWithValue("?pe", Biblioteca.Encripta(EditLiberado.Text));

                    comando.Parameters.AddWithValue("?pf", Biblioteca.MD5String(EditMes.Text + EditAno.Text + EditVencimento.Text + EditBloqueio.Text + EditLiberado.Text));

                    comando.ExecuteNonQuery();
                }
                catch (Exception eError)
                {
                    MessageBox.Show(eError.Message);
                }
                finally
                {
                    if (leitor != null)
                        leitor.Close();
                }
            }
            else
            {
                MessageBox.Show("Não é possível realizar mais lançamentos nessa tabela.");
            }
        }

        private void Button2_Click(object sender, System.EventArgs e)
        {
            ConsultaSQL = "select * from controle_pagamento";
            try
            {
                conexao = dbConnection.conectar();

                ListaPagamento = new List<ControlePagamentoVO>();

                comando = new MySqlCommand(ConsultaSQL, conexao);
                leitor = comando.ExecuteReader();
                while (leitor.Read())
                {
                    ControlePagamentoVO Pagamento = new ControlePagamentoVO();

                    Pagamento.Id = Convert.ToInt32(leitor["id"]);
                    Pagamento.a = Convert.ToString(leitor["a"]);
                    Pagamento.b = Convert.ToString(leitor["b"]);
                    Pagamento.c = Convert.ToString(leitor["c"]);
                    Pagamento.d = Convert.ToString(leitor["d"]);
                    Pagamento.e = Convert.ToString(leitor["e"]);
                    Pagamento.f = Convert.ToString(leitor["f"]);

                    ListaPagamento.Add(Pagamento);
                }

                Grid1.DataSource = ListaPagamento;
            }
            catch (Exception eError)
            {
                MessageBox.Show(eError.Message);
                ListaPagamento = null;
            }
            finally
            {
                if (leitor != null)
                    leitor.Close();
            }
        }



    }
	
	
}
